src/boot/store.js

```diff
@@ -253,6 +253,16 @@ const migrations: { [string]: (GlobalState) => GlobalState } = {
     ),
   }),
 
+  // Change format of keys representing PM narrows, dropping emails.
+  '21': state => ({
+    ...dropCache(state),
+    drafts: objectFromEntries(
+      Object.keys(state.drafts)
+        .map(key => key.replace(/^pm:d:(.*?):.*/s, 'pm:$1'))
```

Comment:
Good, I think `'21'` is safe from the potential issue mentioned in `'19'`:

```
      // Note this will migrate straight to our current format, even after
      // that changes from when this migration was written!  That saves us
      // from duplicating `keyFromNarrow` here... but calls for care in
      // migrations for future changes to `keyFromNarrow`.
```

In particular:

- If [`'21'`] is the sequence of migrations that run at startup, drafts in `'21'`'s input will be in the `pm:d:` format, which this string replacement handles correctly.
- If [`'20'`, `'21'`] is the sequence of migrations that run at startup, drafts in `'21'`'s input will be...empty. (Drafts in `'20'`'s input would have all had `pm:s:` keys, and `'20'` would have removed them all.) Drafts being empty in `'21'`'s input is fine; it just gives an empty output.
- If [..., `'19'`, `'20'`, `'21'`] is the sequence of migrations that run at startup, drafts in `'21'`'s input will have keys that are already in the up-to-date `pm:` format (because of the quirk with `'19'`). This is fine; `'21'` is idempotent on those keys because the `.replace` neither throws an error with that input nor changes it at all.

Reply:
Thanks for checking this!

> * If [`'20'`, `'21'`] is the sequence of migrations that run at startup, drafts in `'21'`'s input will be...empty.

(It'll have no PMs, that is -- it'll still have drafts for stream messages.)


> * If [(...,) `'19'`, `'20'`, `'21'`] is the sequence of migrations that run at startup, drafts in `'21'`'s input will have keys that are already in the up-to-date `pm:` format

Hmm. I think there's actually a bug here, though! In migration 19:
```js
      Object.keys(state.drafts).map(key => [keyFromNarrow(JSON.parse(key)), state.drafts[key]]),
```
the `JSON.parse` for a PM narrow will return something like `{ operator: 'pm-with', operand: 'foo@example.com,bar@example.org' }`. In fact this is really two bugs:
* That's the old shape of `Narrow`, which `keyFromNarrow` no longer accepts.
* That has emails and no user IDs. So there's nothing `keyFromNarrow` could possibly do to convert it to the new format.

The bug was introduced with #4346, where `Narrow` changed and so the type expected by `keyFromNarrow` changed. The reason Flow didn't catch that there was this hole in the encapsulation is that `JSON.parse` there.

This probably means that my plan in #4339 (at https://github.com/zulip/zulip-mobile/pull/4339#discussion_r543603022 ) wasn't the right one in retrospect. It would have been better to simply drop the drafts, or else bite the bullet and duplicate `keyFromNarrow`.

Happily no general users can have been affected at all, because we haven't made a release since this was introduced.

I think I'll fix this by rolling up migrations 19 and 20 into a new migration 21, which just drops drafts entirely (plus `dropCache`). That'll have the same effect in practice as if I'd gone with a "drop drafts" plan in #4339 from the beginning -- each user will have drafts dropped once. And the current migration 20 is already meant to drop all PM drafts, so only stream drafts are newly affected.

Reply:
Oh OK, great—I'm glad you found this! 🙂

> I think I'll fix this by rolling up migrations 19 and 20 into a new migration 21

SGTM.